sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/GlobalEndpointManager.java

```diff
@@ -253,8 +274,27 @@ private void startRefreshLocationTimerAsync() {
     }
 
     private Mono<DatabaseAccount> getDatabaseAccountAsync(URI serviceEndpoint) {
-        return this.owner.getDatabaseAccountFromEndpoint(serviceEndpoint)
-            .doOnNext(i -> logger.debug("account retrieved: {}", i)).single();
+        final GlobalEndpointManager that = this;
+        Callable<Mono<DatabaseAccount>> fetchDatabaseAccount = () -> {
+            return that.owner.getDatabaseAccountFromEndpoint(serviceEndpoint).doOnNext(databaseAccount -> {
+                if(databaseAccount != null) {
+                    this.latestDatabaseAccount = databaseAccount;
+                }
+
+                logger.debug("account retrieved: {}", databaseAccount);
+            }).single();
+        };
+
+        Mono<DatabaseAccount> obsoleteValueMono = databaseAccountAsyncCache.getAsync(StringUtils.EMPTY, null, fetchDatabaseAccount);
+        return obsoleteValueMono.flatMap(obsoleteValue -> {
+            if (firstTimeDatabaseAccountInitialization.compareAndSet(true, false)) {
+                return Mono.just(obsoleteValue);
+            }
+            return databaseAccountAsyncCache.getAsync(StringUtils.EMPTY, obsoleteValue, fetchDatabaseAccount).doOnError(t -> {
+                //Putting back the old value in cache, this will avoid cache corruption
+                databaseAccountAsyncCache.set(StringUtils.EMPTY, obsoleteValue);
```

Comment:
does this mean there is a window of time where there is an invalid result in the cache?

Reply:
Yes

Reply:
Removed cache as per offline discussion